src/ratelimit_manager.cpp

```diff
@@ -0,0 +1,344 @@
+    // lock mutex
+    std::unique_lock<std::shared_mutex>lock(rate_limit_mutex);
+
+    for(const auto &entity : entities) {
+        if (rate_limit_entities.count(rate_limit_resource_t{entity_type,entity}) > 0) {
+            return false;
+        }
+    }
+
+    // add rate limit for entity
+    rule_store.insert({last_rule_id,rate_limit_rule_t{last_rule_id,RateLimitAction::THROTTLE, entity_type,entities, rate_limit_max_requests_t{minute_threshold, hour_threshold}, auto_ban_threshold_num, auto_ban_num_days}});
+
+    // Add rule from rule store to rate limit rule pointer
+    for(const auto &entity : entities) {
+        rate_limit_entities.insert({rate_limit_resource_t{entity_type,entity}, &rule_store.at(last_rule_id)});
+    }
+
+    last_rule_id++;;
+    return true;
+
+}
+
+bool RateLimitManager::remove_rule_entity(const RateLimitedEntityType entity_type, const std::string &entity) {
+    // lock mutex
+    std::unique_lock<std::shared_mutex>lock(rate_limit_mutex);
+
+    // Check if a rule exists for the given IP
+    if (rate_limit_entities.count(rate_limit_resource_t{entity_type,entity}) == 0) {
+        return false;
+    }
+
+    // Remove the rule from the rule store
+    rule_store.erase(rate_limit_entities.at(rate_limit_resource_t{entity_type,entity})->id);
+
+    // Remove the rule from the entity rate limits map
+    rate_limit_entities.erase(rate_limit_resource_t{entity_type,entity});
+
+    return true;
+
+}
+
+void RateLimitManager::temp_ban_entity(const rate_limit_resource_t& entity, const int64_t number_of_days) {
+    // lock mutex
+    std::unique_lock<std::shared_mutex>lock(rate_limit_mutex);
+    temp_ban_entity_wrapped(entity, number_of_days);
+}
+
+bool RateLimitManager::ban_entities_permanently(const RateLimitedEntityType entity_type, const std::vector<std::string> &entities) {
+    // lock mutex
+    std::unique_lock < std::shared_mutex > lock(rate_limit_mutex);
+
+    // Check if a rule exists for the given IP addresses
+    for(const auto& entity : entities) {
+        if (rate_limit_entities.count(rate_limit_resource_t{entity_type,entity}) > 0) {
+            return false;
+        }
+    }
+
+    // add rate limit for entity
+    rule_store.insert({last_rule_id,rate_limit_rule_t{last_rule_id,RateLimitAction::BLOCK, entity_type,entities, rate_limit_max_requests_t{0, 0}, -1, -1}});
+
+    // Add rule from rule store to rate limit rule pointer
+    for(const auto &entity : entities) {
+        rate_limit_entities.insert({rate_limit_resource_t{entity_type,entity}, &rule_store.at(last_rule_id)});
+    }
+
+    last_rule_id++;
+    return true;
+}
+
+
+
+bool RateLimitManager::is_rate_limited(const std::vector<rate_limit_resource_t> &entities) {
+    // lock mutex
+    std::unique_lock<std::shared_mutex>lock(rate_limit_mutex);
+    
+    // Check if any of the entities has rule
+    for(const auto &entity : entities) {
+
+        if (rate_limit_entities.count(entity) == 0) {
```

Comment:
How will you handle wildcard `.*` rules here? For e.g. rule for any IP or any API key will use `.*` for IP address or API key.